panoptes_aggregation/reducers/polygon_reducer_utils.py

```diff
@@ -0,0 +1,449 @@
+    geo2 = data_in[int(b[0])]['polygon']
+
+    # The two shapes need to intersect and be simple for the intersection area
+    # to be found
+    if geo1.intersects(geo2) and (shapely.is_simple(geo1) and shapely.is_simple(geo2)):
+        intersection = geo1.intersection(geo2).area
+    else:  # No intersection, so distance is 1
+        return 1
+
+    union = geo1.union(geo2).area
+
+    return 1 - intersection / union
+
+
+def IoU_distance_matrix_of_cluster(cdx, X, data):
+    '''Find distance matrix using `IoU_metric_polygon` for a cluster.
+
+    The `cdx` argument is used to define the cluster out of the full `X` and
+    `data` data sets, which may also contain other polygons not in the cluster.
+
+    Parameters
+    ----------
+    cdx : numpy.ndarray
+        A 1D array of booleans, corresponding to the polygons in `X` and `data`
+        which are in the cluster. `True` if in the cluster, `False` otherwise.
+    X : numpy.ndarray
+        A 2D array with each row mapping to the data held in `data`. The first
+        column contains row indices and the second column is an index assigned
+        to each user.
+    data_cluster : list
+        A list of dicts that take the form
+        {`polygon`: shapely.geometry.polygon.Polygon, 'time': float, 'gold_standard', bool}
+        There is one element in this list for each member of the cluster.
+
+    Returns
+    -------
+    distances_matrix : numpy.ndarray
+        A symmetric-square array, with the off-diagonal elements containing the
+        IoU distance between the cluster members. The diagonal elements are all
+        zero.
+    '''
+    cluster_X = X[cdx]
+    num_in_cluster = np.shape(cluster_X)[0]
+    # If a cluster of 1 is provided, this will correctly default to a distance of 0
+    distances_matrix = np.zeros((num_in_cluster, num_in_cluster))
+    for i in range(num_in_cluster):
+        a = cluster_X[i]
+        for j in range(i + 1, num_in_cluster):
+            b = cluster_X[j]
+            distance = IoU_metric_polygon(a, b, data_in=data)
+            distances_matrix[i, j] = distance
+            distances_matrix[j, i] = distance
+    return distances_matrix
+
+
+def IoU_cluster_mean_distance(distances_matrix):
+    '''The mean `IoU_metric_polygon` distance between the polygons of the
+    cluster.
+
+    Parameters
+    ----------
+    distances_matrix : numpy.ndarray
+        A symmetric-square array, with the off-diagonal elements containing the
+        `IoU_metric_polygon` distance between the cluster members. The diagonal
+        elements are all zero. This is found using `IoU_distance_matrix_of_cluster`.
+
+    Returns
+    -------
+    distances_mean : float
+        The mean of the `IoU_metric_polygon` defined distance between the
+        polygons of the cluster.
+    '''
+    # first check if it is symmetric
+    if not issymmetric(distances_matrix):
+        raise Exception('`distances_matrix` must be a symmetric-square array')
```

Comment:
Do we really need to check this?  We construct that distance matrix above and know it is symmetric, or am I missing something?